### DISCUSSION TITLE
How to correctly publish an extension?

### DISCUSSION THREAD

**Discussion by pftnhr** [datenstrom/community#892] [Ask a question] [answered]
I've made a little feature, which I'd like to publish. After a while I had all folders `publish.php` wanted me to create and one appeal worked. After that I made some changes und ran the command again. The answer from my terminal is:
``` bash
❯ php yellow.php publish yellow-headinglink
Publishing extension files 100%... done
Yellow publish: 1 extension, 0 errors
```
but nothing happened in the `yellow-extensions` folder. Even when I deleted the `downloads` folder, I wasn't notified that it was missing.

What am I doing wrong?

**Comment by pftnhr** [comment]
PHP 8.2.9 is installed and publish still doesn't work.

**Comment by schulle4u** [comment]
As long as you have configured `Status: experimental` in the [installation settings](https://github.com/pftnhr/yellow-headinglink/blob/main/extension.ini) your extension won't show up in the official repository. The publish command however should update the version number and publication date in your own repository. 

Se also: [Supported status values](https://github.com/annaesvensson/yellow-publish#settings-status).

**Reply by pftnhr** [reply]
Unfortunately, nothing happens in my repository. No publication date is changed, no zip is created (which doesn't bother me).

I have now deleted the status from the extension.ini and then a zip was created. Besides the extension, it only contains the ini and no readme, no screenshot, nothing.

Is there a detailed point-by-point guide for creating extensions somewhere?

**Reply by schulle4u** [reply]
Only the files you have defined in the installation settings are inside the zip file. Screenshots and documentation files are generally not needed anywhere else, therefore they are left out.

**Reply by pftnhr** [reply]
Ok, I was not aware of that. What remains is the fact that the date in the `extension.ini` isn't updated.

**Reply by pftnhr** [reply]
I could care less about that. I'm only trying this to test whether it's because of the publish process that my extension doesn't write language defaults to the `language.ini` (https://github.com/datenstrom/yellow/discussions/898).

**Reply by schulle4u** [reply]
The publication date will be updated when you raise the version number.

**Reply by pftnhr** [reply]
Ok, I'll a look on that. Thank you!

**Comment by annaesvensson** [comment]
The specific problem with the yellow-headinglink repository should have been solved by Steffen's answer. However the output of the [publish command](https://github.com/annaesvensson/yellow-publish) was really confusing in this case, you could also say it was completely wrong, it said that 1 extension was published despite the experimental status in the settings. As a rule of thumb we want to remove complexity from people's everyday lives and this doesn't strike me as particularly straightforward.

I would like to get more feedback about making extensions from developers and designers. Maybe the existing tools are still too confusing and the existing documentation doesn't give enough explanation for beginners. Someone mentioned that a point-by-point guide is missing. What are the problems you encountered during the making of extensions? Can you describe what was missing from your point of view? Do you think we should document one additional step what you can do between "making an extension" and "publishing an extension"?

Let us know what you think. Any comments are welcome.

**Reply by pftnhr** [reply]
I would definitely welcome a point-to-point guide! The shell script does tell you that you have to create a folder with the name of the extension and a folder called `yellow-extensions`, and it also tells you that you have to enter the output path in `yellow-system.ini`.

Personally, however, I originally assumed that the publish script would fetch the extension files from the respective folders and more or less write the `extension.ini` itself. This would, of course, make the procedure a lot easier, as you could then have an installation for each extension in which you can test it extensively without having to constantly copy files back and forth.

But perhaps I am wishing for simplicity, the realisation of which would drive complexity to unimaginable heights.

However, I also imagine a point-to-point manual to be very complex when I think of how my setup looks:

* Server: MAMP Pro with PHP 8.2
* Environment: macOS with PHP 8.2 (homebrew)

but maybe that's not so tragic, since the productive server of most people is usually a web hoster.

If you could explain how the publish extension works, i.e. which files and folders have to be available and where, that would certainly help some people.

**Reply by annaesvensson** [reply, marked as answer]
Now there's a new way of publishing extensions. The [publish command](https://github.com/annaesvensson/yellow-publish) will take your meta information and modify the file `system/extensions/update-available.ini` in [the main repository](https://raw.githubusercontent.com/datenstrom/yellow/main/system/extensions/update-available.ini). This is the central place where we store the update settings for available extensions. Previously extensions were stored in the repository `datenstrom/yellow-extensions`, but this will be no longer needed in favour of individual repositories. There are a few more things the publish command can do, but this is done more or less automatically to make your life easier, for example the publish command can extract version number and publication date from your source code.

Here's what you do if you want to publish a new version of your extension:

1.  Increase the version number in your source code, for example `const VERSION = "1.2.3"`.
2. Copy the source code into your local GitHub folder, for example `/Users/anna/Documents/GitHub/yellow-stockholm/`.
3. Run the publish command at the command line, for example `php yellow.php publish yellow-stockholm`.
4. Upload your changes to GitHub and create a pull request for the repository `datenstrom/yellow`.

Hope this is less complicated.

**Reply by pftnhr** [reply]
That certainly sounds like it. I'll report back as soon as I create or update another extension.

**Reply by annaesvensson** [reply]
How was the experience with the latest publish command?

**Reply by pftnhr** [reply]
I still have to make sure that I am in the right directory and that all paths are correct. In the `extension.ini` I also have to update the published date manually each time.

I wish there was a `php yellow.php dev installation` prompt, which would load the publish extension, create a directory for the things to be published and write the path to the `yellow-system.ini`. Then you would have everything you need with one installation.

**Reply by GiovanniSalmeri** [reply]
I drop a line just to say that I couldn't yet test the new publish command, but I appreciate very much that a simpler way of publishing has been developed! I will come back with my impressions as soon as possible 😊

**Reply by annaesvensson** [reply]
> I also have to update the published date manually each time.

Check that you have increased the version number in the source code before publishing an extension.

**Reply by pftnhr** [reply]
Ah! That makes sense... 🙄😄

**Reply by annaesvensson** [reply]
Thank you, I added this point to the self review checklist. Unfortunately you can not rely on the file date alone, because repositories do not save the original file date, so every time you check out or revert source code, the file date changes, which could result in many unintended changes. Maybe not that bad with only one extension, but not what you want if you maintain multiple extensions. The publish command checks the version number in your source code, then it knows with certainty when you do want to publish a new version and takes the current file date as the publishing date.

**Reply by markseuffert** [reply]
>  I appreciate very much that a simpler way of publishing has been developed!

Did you have time to test the new publish command, how did it go? 🤔

**Reply by GiovanniSalmeri** [reply]
Finally I have found time and serenity of mind for learning the new simplified process of publishing! All seems to me very easy. Perhaps a slightly more detailed guide would be useful for newbies like me (for example, I struggled a little to understand what the setting `PublishWebsiteDirectory` was). Now I have about forty extensions eagerly awaiting the publication 😁

Edit: I'm afraid I did somehing wrong. Before issuing the command `publish` I deleted the setting `Published` in `extension.ini` because I had unterstood that it would have been automatically created... But now I realise that it's not like this, I am a bit confused....

**Reply by markseuffert** [reply]
It sounds like there's room for further improvement. The publish command was designed to modify existing extension settings and files. I think it's kind of expected to have a `Published` setting in all `extension.ini` files. Not having a publication date in extension settings sounds like an edge case, this should never happen. Any suggestions what we should do? Maybe show an error message?

**Reply by GiovanniSalmeri** [reply]
My misunderstanding came from the terminology `publish.php`, `Status: public`, `Published: ####-##-## ##:##:##` (for an Italian-thinking person, "publish" [*pubblicare*] means exactly "make public" [*pubblico*]). I thought: well, if an extension is not yet "published", then it must not have a "published" time... Perhaps `Status: official` and/or `Updated: ####-##-## ##:##:##` or something like this would be clearer?

Moreover, I added to `exension.ini` the setting `Status: public` because I didn't know what the default value was and I imagined that for security reasons it was `experimental` (in this way a command `php yellow.php publish all` wouldn't accidentally publish unwanted extensions). Now I understand that the default value is `public`, am I right?

Anyway, a message error would do. Or else the simple logic that I wrongly supposed (if `Published` is not present, then add it) could be implemented, perhaps it's better than an error message from the point of view of usability?

In any case: the new way of publishing extensions is really very good! 👍

**Reply by markseuffert** [reply]
Good idea, the code has been updated as suggested. `Published` and `Status` are now mandatory settings.

**Comment by schulle4u** [comment]
For the curious, I've published my current [local Yellow development environment](https://gist.github.com/schulle4u/d970e9060d47e768f15f37e79611fe24). I am a Windows user, but perhaps this might help as an example to close the gap between developing and publishing an extension.

**Reply by pftnhr** [reply]
Wow Steffen, thank you so much for this! That already helps with the set-up. But a few question marks remain:

1. in which directory do you develop and test your extensions?
2. how do your extension files get into the directory `/GitHub/yellow-name`?

Maybe there are more questions but these are the ones that interest me the most.

Thank you in advance
Robert

**Reply by schulle4u** [reply]
This is easier than you think. When building a new extension I already use the `yellow-name` directory as project folder and once I'm ready to publish my code, I simply initialize a new GitHub repository in this folder. The testing part however is still a little less comfortable, I copy my files by hand over into a Yellow test installation running in my WSL. But since a successful Yellow setup only takes two commands this is an acceptable  effort for me. You could even use the same Yellow setup for publishing to test your extensions, but I personally like to keep things clean and mostly use a separate installation for testing purposes.
